### PR TITLE
[R-package][docs] made roxygen2 tags explicit and cleaned up documentation

### PR DESCRIPTION
**Pull request by jameslamb**
We use `roxygen2` to document the R package. You can read more about this project [here](http://r-pkgs.had.co.nz/man.html), but basically you use specially-formatted comments in source code files and they get translated into R's special flavor of LaTeX (`.Rd`).

Right now, `lightgbm`'s roxygen2 stuff is relying on the  _convention_ that the first of these comments maps to the `title` element of a documentation page and the the first of these comments after that (broken by a blank line) becomes the `description` element. This can be error-prone, so in the interest of being explicit in this PR I propose that we _explicitly_ use the `@title` and `@description` tags.

Other changes in this PR:

* added previously-undeclared import of `utils`
* removed redundant uses of `@rdname` and replaced them with `@name`
* added docs on the expected form of `group` when using  `Dataset$setinfo("group")`
* clarified documentation on the expected form of `categorical_feature`

This PR also fixes LaTeX formatting issues that were smashing parameter names into their descriptions like this:

**Before**

lgb train-params-before

**After**

lgb train-params-after

### CODE REVIEW

**Comment by StrikerRUS** [comment]
@jameslamb Please merge this PR to a new branch in LightGBM's repo. It will allow to activate RTD builds for that branch and inspect proposed changes visually at rendered documents.

**Comment by jameslamb** [comment]
> @jameslamb Please merge this PR to a new branch in LightGBM's repo. It will allow to activate RTD builds for that branch and inspect proposed changes visually at rendered documents.

oh cool, ok I can do that

File: R-package/man/lightgbm.Rd [hunk 1]
```diff
@@ -43,26 +43,27 @@ If early stopping occurs, the model will have 'best_iter' field.}
...
-   \item{eval}{evaluation function, can be (a list of) character or custom eval function}
-   \item{record}{Boolean, TRUE will record iteration message to \code{booster$record_evals}}
-   \item{colnames}{feature names, if not null, will use this to overwrite the names in dataset}
-   \item{categorical_feature}{list of str or int. type int represents index, type str represents feature names}
-   \item{reset_data}{Boolean, setting it to TRUE (not the default value) will transform the booster model
+   \item{\code{eval}: evaluation function, can be (a list of) character or custom eval function}
+   \item{\code{record}: Boolean, TRUE will record iteration message to \code{booster$record_evals}}
+   \item{\code{colnames}: feature names, if not null, will use this to overwrite the names in dataset}
+   \item{\code{categorical_feature}: categorical features. This can either be a character vector of feature
+                       names or an integer vector with the indices of the features (e.g. \code{c(1L, 1L)} to
```
Comment (StrikerRUS) [inline R-package/man/lightgbm.Rd]: I think `.Rd` files should be updated manually after the recent fixes.

**Comment by StrikerRUS** [comment]
@jameslamb I switched on builds for the `test_roxygen` branch. Now you should be able to see them.

**Comment by StrikerRUS** [comment]
Right now all builds encounter the following error:
```
/bin/bash: line 4:   850 Killed                  /home/docs/.conda/bin/conda install -q -y -n r_env -c conda-forge r-pkgdown=1.3.0=r351h6115d3f_1000 r-roxygen2=6.1.1=r35h0357c0b_1001
Error in library(pkgdown) : there is no package called ‘pkgdown’
Execution halted
```

I remember this error already happened two days ago. It disappeared after a few hours without any actions from us.

**Comment by StrikerRUS** [comment]
Everything is working again now: https://lightgbm.readthedocs.io/en/test_roxygen/R/reference/index.html.

File: R-package/R/lgb.cv.R [hunk 2]
```diff
@@ -36,19 +36,19 @@ CVBooster <- R6::R6Class(
...
+#'                            \code{c(1L, 10L)} to say "the first and tenth columns").
 #' @param callbacks List of callback functions that are applied at each iteration.
 #' @param reset_data Boolean, setting it to TRUE (not the default value) will transform the booster model
 #'                   into a predictor model which frees up memory and the original datasets
 #' @param ... other parameters, see Parameters.rst for more information. A few key parameters:
 #'            \itemize{
-#'                \item{boosting}{Boosting type. \code{"gbdt"} or \code{"dart"}}
-#'                \item{num_leaves}{number of leaves in one tree. defaults to 127}
-#'                \item{max_depth}{Limit the max depth for tree model. This is used to deal with
+#'                \item{\code{boosting}: Boosting type. \code{"gbdt"} or \code{"dart"}}
```
Comment (StrikerRUS) [inline R-package/R/lgb.cv.R]:
```suggestion
#'                \item{\code{boosting}: Boosting type. \code{"gbdt"}, \code{"rf"}, \code{"dart"} or \code{"goss"}.}
```

File: R-package/R/lgb.cv.R [hunk 3]
```diff
@@ -36,19 +36,19 @@ CVBooster <- R6::R6Class(
...
+#'                \item{\code{num_leaves}: number of leaves in one tree. defaults to 127}
```
Comment (StrikerRUS) [inline R-package/R/lgb.cv.R]: From where this default value is taken from? If from cpp side, then it is not true anymore: https://lightgbm.readthedocs.io/en/latest/Parameters.html#num_leaves
```suggestion
#'                \item{\code{num_leaves}: Number of leaves in one tree. Defaults to 127.}
```

File: R-package/R/lgb.prepare2.R [hunk 4]
```diff
@@ -1,14 +1,12 @@
...
-#' Please use \code{lgb.prepare_rules2} if you want to apply this transformation to other datasets.
-#' This is useful if you have a specific need for integer dataset instead of numeric dataset.
-#' Note that there are programs which do not support integer-only input. Consider this as a half
-#' memory technique which is dangerous, especially for LightGBM.
-#'
+#' @name lgb.prepare2
+#' @title Data preparator for LightGBM datasets (integer)
+#' @description Attempts to prepare a clean dataset to prepare to put in a \code{lgb.Dataset}.
+#'              Factors and characters are converted to numeric (specifically: integer).
+#'              Please use \code{lgb.prepare_rules2} if you want to apply this transformation to
```
Comment (StrikerRUS) [inline R-package/R/lgb.prepare2.R]:
```suggestion
#'              Please use \code{\link{lgb.prepare_rules2}} if you want to apply this transformation to
```

File: R-package/R/lgb.train.R [hunk 5]
```diff
@@ -18,11 +18,11 @@
 #'                   original datasets
 #' @param ... other parameters, see Parameters.rst for more information. A few key parameters:
 #'            \itemize{
-#'                \item{boosting}{Boosting type. \code{"gbdt"} or \code{"dart"}}
-#'                \item{num_leaves}{number of leaves in one tree. defaults to 127}
-#'                \item{max_depth}{Limit the max depth for tree model. This is used to deal with
+#'                \item{\code{boosting}: Boosting type. \code{"gbdt"} or \code{"dart"}}
```
Comment (StrikerRUS) [inline R-package/R/lgb.train.R]:
```suggestion
#'                \item{\code{boosting}: Boosting type. \code{"gbdt"}, \code{"rf"}, \code{"dart"} or \code{"goss"}.}
```

File: R-package/R/lgb.train.R [hunk 6]
```diff
@@ -18,11 +18,11 @@
...
+#'                \item{\code{num_leaves}: number of leaves in one tree. defaults to 127}
```
Comment (StrikerRUS) [inline R-package/R/lgb.train.R]: From where this default value is taken from? If from cpp side, then it is not true anymore: https://lightgbm.readthedocs.io/en/latest/Parameters.html#num_leaves
```suggestion
#'                \item{\code{num_leaves}: Number of leaves in one tree. Defaults to 127.}
```

File: R-package/R/lightgbm.R [hunk 7]
```diff
@@ -25,21 +24,23 @@ NULL
...
+#'        \item{\code{colnames}: feature names, if not null, will use this to overwrite the names in dataset}
+#'        \item{\code{categorical_feature}: categorical features. This can either be a character vector of feature
+#'                            names or an integer vector with the indices of the features (e.g. \code{c(1L, 10L)} to
+#'                            say "the first and tenth columns").}
+#'        \item{\code{reset_data}: Boolean, setting it to TRUE (not the default value) will transform the booster model
 #'                          into a predictor model which frees up memory and the original datasets}
-#'         \item{boosting}{Boosting type. \code{"gbdt"} or \code{"dart"}}
-#'         \item{num_leaves}{number of leaves in one tree. defaults to 127}
-#'         \item{max_depth}{Limit the max depth for tree model. This is used to deal with
+#'         \item{\code{boosting}: Boosting type. \code{"gbdt"} or \code{"dart"}}
```
Comment (StrikerRUS) [inline R-package/R/lightgbm.R]:
```suggestion
#'         \item{\code{boosting}: Boosting type. \code{"gbdt"}, \code{"rf"}, \code{"dart"} or \code{"goss"}.}
```

File: R-package/R/lightgbm.R [hunk 8]
```diff
@@ -25,21 +24,23 @@ NULL
...
+#'         \item{\code{num_leaves}: number of leaves in one tree. defaults to 127}
```
Comment (StrikerRUS) [inline R-package/R/lightgbm.R]: From where this default value is taken from? If from cpp side, then it is not true anymore: https://lightgbm.readthedocs.io/en/latest/Parameters.html#num_leaves
```suggestion
#'         \item{\code{num_leaves}: Number of leaves in one tree. Defaults to 127.}
```

File: R-package/R/lightgbm.R [hunk 9]
```diff
@@ -115,24 +115,20 @@ lightgbm <- function(data,
...
-#' UCI Machine Learning Repository.
-#'
-#' This data set includes the following fields:
-#'
-#' \itemize{
-#'  \item \code{label} the label for each record
-#'  \item \code{data} a sparse Matrix of \code{dgCMatrix} class, with 126 columns.
-#' }
-#'
+#' @title Test part from Mushroom Data Set
```
Comment (StrikerRUS) [inline R-package/R/lightgbm.R]: Place `@name` field on the top?

File: R-package/R/lightgbm.R [hunk 10]
```diff
@@ -148,13 +144,12 @@ NULL
 #' rows and 126 variables
 NULL
 
-#' Bank Marketing Data Set
-#'
-#' This data set is originally from the Bank Marketing data set,
-#' UCI Machine Learning Repository.
+#' @title Bank Marketing Data Set
```
Comment (StrikerRUS) [inline R-package/R/lightgbm.R]: Place `@name` field on the top?

Comment (jameslamb) [reply R-package/R/lightgbm.R]: yep good catch!

File: R-package/R/lightgbm.R (hunk 8 above)
Comment (jameslamb) [reply R-package/R/lightgbm.R]: 😬 I'll  update it, thank you. Not something that is set on the R side.

Comment (jameslamb) [reply R-package/R/lightgbm.R]: This line was duplicated in `lightgbm()`,  `lgb.train()` and `lgb.cv()`. I updated all three to:

```
Maximum number of leaves in one tree.
```

**Comment by jameslamb** [comment]
> Everything is working again now: https://lightgbm.readthedocs.io/en/test_roxygen/R/reference/index.html.

Hey thanks @StrikerRUS ! Could you leave that branch enabled even after this PR? I'm planning a few other updates to the R documentation and would be great to be able to check there.

File: R-package/R/lightgbm.R (hunk 8 above)
Comment (StrikerRUS) [reply R-package/R/lightgbm.R]: Cool! Thanks!

**Comment by StrikerRUS** [comment]
@jameslamb 
> Could you leave that branch enabled even after this PR? 

Sure! I just made it private. You should still have access to it by original link, but users will not be confused by seeing it in versions selection. Please check that the link works.

**Pull request merged**